### PR TITLE
oc logs must wait for first build from bc with ConfigChange trigger

### LINKED ISSUE #12142: `oc logs bc/x` reports "no builds found" if run quickly after bc creation
The following is a common interaction pattern with oc (the output of new-build even encourages it):
$ oc new-build ..... --name=hello
$ oc logs bc/hello -f

However, in a scripted environment (e.g. bash or Jenkins pipeline), where the interaction occurs with super-human speed, oc logs can fail with: error: no builds found for 'stress' .

Instead, oc logs, when encountering this condition, could inspect the buildconfig for a configchange trigger. If the trigger exists, the operation could wait for a reasonable period for the appearance of a build before giving up.

##### Version
oc v1.5.0-alpha.0+0c89011-92
kubernetes v1.4.0+776c994
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://10.13.137.186:8443
openshift v1.5.0-alpha.0+3b2bbe5
kubernetes v1.4.0+776c994

##### Steps To Reproduce
1. Run this script with oc in the path / logged in. https://gist.github.com/jupierce/588b8ccd9f3d8b0df52811f29ce4f3e9

##### Current Result
oc logs fails with error: no builds found for 'stress' .

##### Expected Result
The script should run indefinitely.

### PR DESCRIPTION
**Pull request by jupierce**
Fixes  #12142
ptal @bparees

### CODE REVIEW

File: pkg/cmd/util/clientcmd/factory.go [hunk 1]
```diff
@@ -467,10 +468,34 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
...
+			err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+				builds.Items = buildapi.FilterBuilds(builds.Items, buildapi.ByBuildConfigPredicate(t.Name))
+				if len(builds.Items) > 0 {
+					return true, nil
+				} else {
+					if hasConfigChangeTrigger {
+						// Wait until timeout before giving up on the first build
+						return false, nil
+					} else {
+						return false, fmt.Errorf("no builds found for %q", t.Name)
```
Comment (bparees) [inline pkg/cmd/util/clientcmd/factory.go]: isn't this error going to get wrapped/eaten by line 495?  I think we want to retain the existing output behavior for cases where there was no build found and no configchangetrigger on the BC.

Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: Why are we waiting in this code?  This code should not be waiting for anything.  If you need to wait, do it on the server.

Comment (bparees) [inline pkg/cmd/util/clientcmd/factory.go]: there is no buildconfig/logs api endpoint that could do the waiting, we're just writing a clever client here.

setting up a watch on builds (which would still amount to the client waiting) seems like overkill for this trivial fix.

Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: If this is something that a naive client would want to do, then we should be fixing this on the server.  I am not in favor of overly complex code here to fix one client.

Comment (bparees) [inline pkg/cmd/util/clientcmd/factory.go]: if that is true then the entire "logsforobject" concept is flawed and should have been done as api endpoints, no?

File: pkg/cmd/util/clientcmd/factory.go [hunk 2]
```diff
@@ -467,10 +468,34 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
...
+						hasConfigChangeTrigger = true
+					}
+				}
+			}
+
+			err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+				builds.Items = buildapi.FilterBuilds(builds.Items, buildapi.ByBuildConfigPredicate(t.Name))
+				if len(builds.Items) > 0 {
+					return true, nil
+				} else {
```
Comment (kargakis) [inline pkg/cmd/util/clientcmd/factory.go]: No need for else

File: pkg/cmd/util/clientcmd/factory.go [hunk 3]
```diff
@@ -467,10 +468,34 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
...
+					if hasConfigChangeTrigger {
+						// Wait until timeout before giving up on the first build
+						return false, nil
+					} else {
```
Comment (kargakis) [inline pkg/cmd/util/clientcmd/factory.go]: No need for else

Comment (jupierce) [inline pkg/cmd/util/clientcmd/factory.go]: Using more compact style in next push.

File: pkg/cmd/util/clientcmd/factory.go (hunk 1 above)
Comment (jupierce) [inline pkg/cmd/util/clientcmd/factory.go]: @bparees I've given the potential outcomes their own error messages in the latest push.

File: pkg/cmd/util/clientcmd/factory.go [hunk 4]
```diff
@@ -467,10 +468,34 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 			if err != nil {
 				return nil, err
 			}
-			builds.Items = buildapi.FilterBuilds(builds.Items, buildapi.ByBuildConfigPredicate(t.Name))
-			if len(builds.Items) == 0 {
-				return nil, fmt.Errorf("no builds found for %q", t.Name)
+
+			hasConfigChangeTrigger := false
+			if t.Spec.Triggers != nil && len(t.Spec.Triggers) > 0 {
```
Comment (jupierce) [inline pkg/cmd/util/clientcmd/factory.go]: Added utility method in buildapi.

File: pkg/build/api/helpers.go [hunk 5]
```diff
@@ -59,3 +60,21 @@ func hasBuildConfigAnnotation(build Build, annotationName, annotationValue strin
...
+// Returns an error if no match trigger is found.
+func FindTriggerPolicy(triggerType BuildTriggerType, config *BuildConfig) (buildTriggers []BuildTriggerPolicy, err error) {
+	err = fmt.Errorf("buildconfig trigger type not found: %v", triggerType)
+	for _, specTrigger := range config.Spec.Triggers {
+		if specTrigger.Type == triggerType {
+			buildTriggers = append(buildTriggers, specTrigger)
+			err = nil
+		}
+	}
+	return buildTriggers, err
```
Comment (bparees) [inline pkg/build/api/helpers.go]: Seems weird to return an error for this.  i would expect the caller to check the array length and if it's zero, they know it wasn't found.  errors should be used for issues that are harder to anticipate imho.

Comment (jupierce) [inline pkg/build/api/helpers.go]: I prefer your methodology. My initial pass was to preserve/unify on the signature I found in webhooks. Will make the change.

**Comment by bparees** [comment]
lgtm [merge]

**Pull request merged**